packages/editor/src/dataviews/actions/delete-permanently.ts

```diff
@@ -0,0 +1,120 @@
+/**
+ * WordPress dependencies
+ */
+import { store as coreStore } from '@wordpress/core-data';
+import { store as noticesStore } from '@wordpress/notices';
+import { __, sprintf } from '@wordpress/i18n';
+import type { Action } from '@wordpress/dataviews';
+
+/**
+ * Internal dependencies
+ */
+import { getItemTitle } from './utils';
+import type { Post } from '../types';
+import {
+	TEMPLATE_POST_TYPE,
+	TEMPLATE_PART_POST_TYPE,
+} from '../../store/constants';
+
+type CoreDataError = {
+	reason: { message?: string };
+};
+
+const permanentlyDeletePostAction: Action< Post > = {
+	id: 'permanently-delete',
+	label: __( 'Permanently delete' ),
+	supportsBulk: true,
+	isEligible( { status, type }: Post ) {
+		const isTemplateOrTemplatePart = [
+			TEMPLATE_POST_TYPE,
+			TEMPLATE_PART_POST_TYPE,
+		].includes( type );
+
+		return ! isTemplateOrTemplatePart && status === 'trash';
+	},
+	async callback( posts, { registry } ) {
+		const { createSuccessNotice, createErrorNotice } =
+			registry.dispatch( noticesStore );
+		const { deleteEntityRecord } = registry.dispatch( coreStore );
+		const promiseResult = await Promise.allSettled(
+			posts.map( ( post ) => {
+				return deleteEntityRecord(
+					'postType',
+					post.type,
+					post.id,
+					{ force: true },
+					{ throwOnError: true }
+				);
+			} )
+		);
+		// If all the promises were fulfilled with success.
+		if ( promiseResult.every( ( { status } ) => status === 'fulfilled' ) ) {
+			let successMessage;
+			if ( promiseResult.length === 1 ) {
+				successMessage = sprintf(
+					/* translators: The posts's title. */
+					__( '"%s" permanently deleted.' ),
+					getItemTitle( posts[ 0 ] )
+				);
+			} else {
+				successMessage = __( 'The posts were permanently deleted.' );
+			}
+			createSuccessNotice( successMessage, {
+				type: 'snackbar',
+				id: 'permanently-delete-post-action',
+			} );
+		} else {
+			// If there was at lease one failure.
+			let errorMessage;
+			// If we were trying to permanently delete a single post.
+			if ( promiseResult.length === 1 ) {
+				const result = promiseResult[ 0 ] as CoreDataError;
+				if ( result.reason?.message ) {
+					errorMessage = result.reason.message;
+				} else {
+					errorMessage = __(
+						'An error occurred while permanently deleting the post.'
+					);
+				}
+				// If we were trying to permanently delete multiple posts
+			} else {
+				const errorMessages = new Set();
+				const failedPromises = promiseResult.filter(
+					( { status } ) => status === 'rejected'
+				);
+				for ( const failedPromise of failedPromises ) {
+					const result = failedPromise as CoreDataError;
```

Comment:
This type cast is not ideal, but I'm not sure what I need to change to avoid it. It comes from CoreData, so it might be challenging to type that properly. cc @sirreal @jsnajdr

Reply:
I confirmed that TypeScript 5.5.0 is going to be much smarter about this code. In [this example](https://www.typescriptlang.org/play/?target=99&ts=5.5.0-beta&ssl=1&ssc=1&pln=7&pc=2#code/IYZwngdgxgBAZgV2gFwJYHsL3egPAFQD4AKABwCd0BbEALgAVKrUQBTAwgbQF0BKGAN4AoGDCiYQyGOVYgYAXhjAA7sFRTG1FqwB0wADb6Ayq2TJ9rACZkmIXgG4RYiVLhr9CGXMVedcVPrIrOTE5AqE0jqSwMgI3vKKAOQyAFasUEGWiQ5OcOhhxOIQkvDunqww6HClAeV2gk6iRSDoFjr66ADmxG61MjoyoJg5ogC+QuNAA):
```js
async function foo<T>(proms:Promise<T>[]) {
  const res = await Promise.allSettled(proms);
  const failures = res.filter(r => r.status === 'rejected');
  for (const failure of failures) {
    console.log(failure.reason);
  }
}
```
the new 5.5.0 version automatically detects that `failures` is a `PromiseRejectedResult[]` and that it's OK to access the `.reason` field. The current 5.4.5 will flag `.reason` as type error because it thinks `failures` is still the more general `PromiseSettledResult<T>[]`.

I'm looking forward to the new version 🙂

Reply:
TypeScript 5.5 has been released so we can see about upgrading. It does bring some nice features.